atom/browser/atom_blob_reader.cc

```diff
@@ -53,22 +53,21 @@ AtomBlobReader::AtomBlobReader(content::ChromeBlobStorageContext* blob_context)
 
 AtomBlobReader::~AtomBlobReader() {}
 
-void AtomBlobReader::StartReading(
-    const std::string& uuid,
-    const AtomBlobReader::CompletionCallback& completion_callback) {
+void AtomBlobReader::StartReading(const std::string& uuid,
+                                  atom::util::Promise promise) {
   DCHECK_CURRENTLY_ON(BrowserThread::IO);
 
   auto blob_data_handle = blob_context_->context()->GetBlobDataFromUUID(uuid);
-  auto callback = base::Bind(&RunCallbackInUI, completion_callback);
   if (!blob_data_handle) {
-    base::PostTaskWithTraits(FROM_HERE, {BrowserThread::UI},
-                             base::BindOnce(callback, nullptr, 0));
+    atom::util::Promise::RejectPromise(std::move(promise),
+                                       "Could not get blob data handle");
     return;
   }
 
   auto blob_reader = blob_data_handle->CreateReader();
-  BlobReadHelper* blob_read_helper =
-      new BlobReadHelper(std::move(blob_reader), callback);
+  BlobReadHelper* blob_read_helper = new BlobReadHelper(
+      std::move(blob_reader),
+      base::Bind(&RunPromiseInUI, atom::util::CopyablePromise(promise)));
```

Comment:
Creating `CopyablePromise` involves V8 APIs internally, so it can not be created on IO thread.

For this PR, I think the best solution is to change `BlobReadHelper::CompletionCallback` to use `base::OnceCallback`, and then use move-only `Promise` in all places.